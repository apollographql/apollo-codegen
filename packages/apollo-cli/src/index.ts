// run the CLI
export { run } from "@oclif/command";
