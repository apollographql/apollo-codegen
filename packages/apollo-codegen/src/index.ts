import 'core-js/fn/object/values';
import 'core-js/fn/object/entries';

export { default as downloadSchema } from './downloadSchema';
export { default as introspectSchema } from './introspectSchema';
export { default as printSchema } from './printSchema';
export { default as generate } from './generate';
