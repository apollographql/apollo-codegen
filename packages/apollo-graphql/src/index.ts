export {
  defaultEngineReportingSignature,
  defaultOperationRegistrySignature,
  operationRegistrySignature,
  operationHash
} from "./operationId";
export * from "./schema";
export {
  printWithReducedWhitespace,
  hideStringAndNumericLiterals,
  hideLiterals
} from "./transforms";
