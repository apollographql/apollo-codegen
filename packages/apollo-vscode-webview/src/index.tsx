import * as React from "react";
import * as ReactDOM from "react-dom";
import App from "./App";

declare function acquireVsCodeApi(): {
  postMessage(message: any): void;
};

export const vscode = acquireVsCodeApi();

ReactDOM.render(<App />, document.getElementById("root") as HTMLElement);
