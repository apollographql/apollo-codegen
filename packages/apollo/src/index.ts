// run the CLI
export { run } from "@oclif/command";

// export utils for usage
export * from "./printer/ast";

export { printFromSchemas, printChanges } from "./printer/print";
export { diffSchemas } from "./printer/diff";
