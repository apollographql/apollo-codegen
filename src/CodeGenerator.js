import {
  join,
  wrap,
} from './utilities/printing';

export default class CodeGenerator {
  constructor() {
    this.scopeStack = [];

    this.indentWidth = 2;
    this.indentLevel = 0;

    this.output = '';
  }

  pushScope(scope) {
    this.scopeStack.push(scope);
  }

  popScope() {
    return this.scopeStack.pop();
  }

  print(maybeString) {
    if (maybeString) {
      this.output += maybeString;
    }
  }

  printNewline() {
    if (this.output) {
      this.print('\n');
      this.startOfIndentLevel = false;
    }
  }

  printNewlineIfNeeded() {
    if (!this.startOfIndentLevelLevel) {
      this.printNewline();
    }
  }

  printOnNewline(maybeString) {
    if (maybeString) {
      this.printNewline();
      this.printIndent();
      this.print(maybeString);
    }
  }

  printIndent() {
    const indentation = ' '.repeat(this.indentLevel * this.indentWidth);
    this.output += indentation;
  }

  withIndent(closure) {
    if (!closure) return;

    this.indentLevel++;
    this.startOfIndentLevel = true;
    closure();
    this.indentLevel--;
  }

  withinBlock(closure) {
    this.print(' {');
    this.withIndent(closure);
    this.printOnNewline('}');
  }
}
