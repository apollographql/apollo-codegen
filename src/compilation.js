import {
  print,
  visit,
  visitWithTypeInfo,
  typeFromAST,
  getNamedType,
  isAbstractType,
  isTypeSubTypeOf,
  Kind,
  TypeInfo,
  isType,
  isCompositeType,
  GraphQLEnumType,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType,
  GraphQLError
} from 'graphql';

import {
  getOperationRootType,
  getFieldDef
} from './utilities/graphql';

import {
  join,
  block,
  wrap,
  indent
} from './utilities/printing';

// Parts of this code are adapted from graphql-js

export class Compiler {
  constructor(schema, document) {
    this.schema = schema;

    this.typesUsedSet = new Set();

    this.fragmentMap = Object.create(null);
    this.operations = [];

    for (const definition of document.definitions) {
      switch (definition.kind) {
        case Kind.OPERATION_DEFINITION:
          this.operations.push(definition);
          break;
        case Kind.FRAGMENT_DEFINITION:
          this.fragmentMap[definition.name.value] = definition;
          break;
      }
    }
  }

  get typesUsed() {
    return Array.from(this.typesUsedSet);
  }

  fragmentNamed(fragmentName) {
    return this.fragmentMap[fragmentName];
  }

  get fragments() {
    return Object.values(this.fragmentMap);
  }

  compileOperations() {
    return this.operations.map(this.compileOperation, this);
  }

  compileOperation(operationDefinition) {
    const operationName = operationDefinition.name.value;

    const variables = operationDefinition.variableDefinitions.map(node => {
      const name = node.variable.name.value;
      const type = typeFromAST(this.schema, node.type);
      this.typesUsedSet.add(type);
      return { name, type };
    });

    const source = print(withTypenameFieldAddedWhereNeeded(this.schema, operationDefinition));

    const rootType = getOperationRootType(this.schema, operationDefinition);
    const groupedFieldSet = this.collectFields(rootType, operationDefinition.selectionSet);
    const fragmentsReferencedSet = Object.create(null);
    const fields = this.resolveFields(rootType, groupedFieldSet, fragmentsReferencedSet);

    return { operationName, variables, source, fields, fragmentsReferenced: Object.keys(fragmentsReferencedSet) };
  }

  compileFragments() {
    return this.fragments.map(this.compileFragment, this);
  }

  compileFragment(fragmentDefinition) {
    const fragmentName = fragmentDefinition.name.value;

    const source = print(withTypenameFieldAddedWhereNeeded(this.schema, fragmentDefinition));

    const fragmentType = typeFromAST(this.schema, fragmentDefinition.typeCondition);
    const groupedFieldSet = this.collectFields(fragmentType, fragmentDefinition.selectionSet);
    const fields = this.resolveFields(fragmentType, groupedFieldSet);
    const typeConditions = this.resolveSubTypes(fragmentType, groupedFieldSet);

    return { fragmentName, source, fields, typeConditions };
  }

  collectFields(parentType, selectionSet, groupedFieldSet = Object.create(null), visitedFragmentSet = Object.create(null)) {
    if (!isCompositeType(parentType)) {
      throw new Error(`parentType should be a composite type, but is "${String(parentType)}"`);
    }

    for (const selection of selectionSet.selections) {
      switch (selection.kind) {
        case Kind.FIELD: {
          const fieldName = selection.name.value;
          const responseName = selection.alias ? selection.alias.value : fieldName;

          const field = getFieldDef(this.schema, parentType, selection);
          if (!field) {
            throw new GraphQLError(`Cannot query field "${fieldName}" on type "${String(parentType)}"`, [selection]);
          }
          const fieldType = field.type;

          if (!groupedFieldSet[responseName]) {
            groupedFieldSet[responseName] = [];
          }

          groupedFieldSet[responseName].push([parentType, { ...selection, type: fieldType }]);
          break;
        }
        case Kind.INLINE_FRAGMENT: {
          const typeCondition = selection.typeCondition;
          const inlineFragmentType = typeCondition ?
            typeFromAST(this.schema, typeCondition) :
            parentType;

          this.collectFields(
            isTypeSubTypeOf(this.schema, inlineFragmentType, parentType) ? inlineFragmentType : parentType,
            selection.selectionSet,
            groupedFieldSet,
            visitedFragmentSet
          );
          break;
        }
        case Kind.FRAGMENT_SPREAD: {
          const fragmentName = selection.name.value;

          const fragment = this.fragmentNamed(fragmentName);
          if (!fragment) continue;

          const typeCondition = fragment.typeCondition;
          const fragmentType = typeFromAST(this.schema, typeCondition)

          const effectiveType = isTypeSubTypeOf(this.schema, fragmentType, parentType) ? fragmentType : parentType;

          if (!visitedFragmentSet[effectiveType]) {
            visitedFragmentSet[effectiveType] = {};
          }

          if (visitedFragmentSet[effectiveType][fragmentName]) continue;
          visitedFragmentSet[effectiveType][fragmentName] = true;

          this.collectFields(
            effectiveType,
            fragment.selectionSet,
            groupedFieldSet,
            visitedFragmentSet
          );
          break;
        }
      }
    }
    return groupedFieldSet;
  }

  mergeSelectionSets(parentType, fieldSet, visitedFragmentSet) {
    const groupedFieldSet = Object.create(null);

    for (const [,field] of fieldSet) {
      const selectionSet = field.selectionSet;

      if (selectionSet) {
        this.collectFields(parentType, selectionSet, groupedFieldSet, visitedFragmentSet);
      }
    }

    return groupedFieldSet;
  }

  resolveFields(parentType, groupedFieldSet, fragmentsReferencedSet) {
    const fields = [];

    for (let [fieldName, fieldSet] of Object.entries(groupedFieldSet)) {
      fieldSet = fieldSet.filter(([typeCondition,]) => isTypeSubTypeOf(this.schema, parentType, typeCondition));
      if (fieldSet.length < 1) continue;

      const [,firstField] = fieldSet[0];
      const fieldType = firstField.type;

      const field = { name: fieldName, type: fieldType };

      const unmodifiedFieldType = getNamedType(fieldType);

      if (isCompositeType(unmodifiedFieldType)) {
        const visitedFragmentSet = Object.create(null);
        const subSelectionSet = this.mergeSelectionSets(unmodifiedFieldType, fieldSet, visitedFragmentSet);

        field.fragmentSpreads = fragmentSpreadsForType(unmodifiedFieldType, visitedFragmentSet);

        if (fragmentsReferencedSet) {
          Object.assign(fragmentsReferencedSet, ...Object.values(visitedFragmentSet));
        }

        field.fields = this.resolveFields(unmodifiedFieldType, subSelectionSet, fragmentsReferencedSet);
        field.typeConditions = this.resolveSubTypes(unmodifiedFieldType, subSelectionSet, visitedFragmentSet, fragmentsReferencedSet);
      }

      fields.push(field);
    }

    return fields;
  }

  resolveSubTypes(parentType, groupedFieldSet, visitedFragmentSet = Object.create(null), fragmentsReferencedSet) {
    return this.collectSubTypes(parentType, groupedFieldSet).map(type => {
      const fields = this.resolveFields(type, groupedFieldSet, fragmentsReferencedSet);
      const fragmentSpreads = fragmentSpreadsForType(type, visitedFragmentSet);
      return { type, fields, fragmentSpreads };
    });
  }

  collectSubTypes(parentType, groupedFieldSet) {
    const typeConditions = new Set();
    for (const fieldSet of Object.values(groupedFieldSet)) {
      for (const [typeCondition,] of fieldSet) {
        if (!isTypeSubTypeOf(this.schema, parentType, typeCondition)) {
          typeConditions.add(typeCondition);
        }
      }
    }
    return Array.from(typeConditions);
  }
}

function fragmentSpreadsForType(type, visitedFragmentSet) {
  return visitedFragmentSet[type] ? Object.keys(visitedFragmentSet[type]) : [];
}

const typenameField = { kind: Kind.FIELD, name: { kind: Kind.NAME, value: '__typename' } };

function withTypenameFieldAddedWhereNeeded(schema, ast) {
  const typeInfo = new TypeInfo(schema);

  return visit(ast, visitWithTypeInfo(typeInfo, {
    leave: {
      SelectionSet: node => {
        const parentType = typeInfo.getParentType();

        if (isAbstractType(parentType)) {
          return { ...node, selections: [typenameField, ...node.selections] };
        }
      }
    }
  }));
}

function sourceAt(location) {
  return location.source.body.slice(location.start, location.end);
}

export function stringifyIR(ast, space) {
  return JSON.stringify(ast, function(key, value) {
    if (isType(value)) {
      return String(value);
    }
    return value;
  }, space);
}

export function printIR({ fields, typeConditions, fragmentSpreads }) {
  return fields && wrap('<', join(fragmentSpreads, ', '), '> ')
    + block(fields.map(field =>
      `${field.name}: ${String(field.type)}` + wrap(' ', printIR(field))
    ).concat(typeConditions && typeConditions.map(typeCondition =>
      `${String(typeCondition.type)}` + wrap(' ', printIR(typeCondition)))));
}
