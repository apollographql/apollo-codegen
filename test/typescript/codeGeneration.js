import { stripIndent } from 'common-tags';

import {
  parse,
  isType,
  GraphQLID,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull
} from 'graphql';

import {
  generateSource
} from '../../src/typescript/codeGeneration';

import { loadSchema } from '../../src/loading';
const schema = loadSchema(require.resolve('../starwars/schema.json'));

import CodeGenerator from '../../src/utilities/CodeGenerator';

import { compileToIR } from '../../src/compilation';

describe('TypeScript code generation', function() {
  let generator;
  let compileFromSource;
  let addFragment;

  beforeEach(function() {
    const context = {
      schema: schema,
      operations: {},
      fragments: {},
      typesUsed: {}
    }

    generator = new CodeGenerator(context);

    compileFromSource = (source) => {
      const document = parse(source);
      const context = compileToIR(schema, document);
      generator.context = context;
      return context;
    };

    addFragment = (fragment) => {
      generator.context.fragments[fragment.fragmentName] = fragment;
    };
  });

  describe('#generateSource()', function() {
    test(`should generate simple query operations`, function() {
      const context = compileFromSource(`
        query HeroName {
          hero {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate simple query operations including input variables`, function() {
      const context = compileFromSource(`
        query HeroName($episode: Episode) {
          hero(episode: $episode) {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate simple nested query operations including input variables`, function() {
      const context = compileFromSource(`
        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              name
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate fragmented query operations`, function() {
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero {
            name
            ...heroFriends
          }
        }

        fragment heroFriends on Character {
          friends {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should handle multi-fragmented query operations`, function() {
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero {
            name
            ...heroFriends
            ...heroAppears
          }
        }

        fragment heroFriends on Character {
          friends {
            name
          }
        }

        fragment heroAppears on Character {
          appearsIn
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate query operations with inline fragments`, function() {
      const context = compileFromSource(`
        query HeroAndDetails {
          hero {
            name
            ...HeroDetails
          }
        }

        fragment HeroDetails on Character {
          ... on Droid {
            primaryFunction
          }
          ... on Human {
            height
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate mutation operations with complex input types`, function() {
      const context = compileFromSource(`
        mutation ReviewMovie($episode: Episode, $review: ReviewInput) {
          createReview(episode: $episode, review: $review) {
            stars
            commentary
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate correct typedefs with a single custom fragment`, function() {
      const context = compileFromSource(`
        fragment Friend on Character {
          name
        }

        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              ...Friend
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate correct typedefs with a multiple custom fragments`, function() {
      const context = compileFromSource(`
        fragment Friend on Character {
          name
        }

        fragment Person on Character {
          name
        }

        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              ...Friend
              ...Person
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should handle complex fragments with type aliases`, function() {
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero(episode: NEWHOPE) {
            name
            ...Something
          }
          empireHero: hero(episode: EMPIRE) {
            name
            ...Something
          }
        }

        fragment Something on Character {
          ... on Human {
            friends {
              ... on Human {
                homePlanet
              }

              ... on Droid {
                primaryFunction
              }
            }
          }

          ... on Droid {
            appearsIn
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should correctly handle fragments on interfaces', function() {
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            ... on Human {
              homePlanet
            }

            ... on Droid {
              primaryFunction
            }
          }
        }
      `
      );

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should correctly handle nested fragments on interfaces', function() {
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            friendsConnection {
              friends {
                ...CharacterFragment
              }
            }
          }
        }

        fragment CharacterFragment on Character {
          name
          __typename

          ... on Human {
            homePlanet
          }

          ... on Droid {
            primaryFunction
          }
        }
      `
      );

      const source = generateSource(context);

      expect(source).toMatchSnapshot();
    });

    test('should correctly add typename to nested fragments on interfaces if addTypename is true', function() {
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            friendsConnection {
              friends {
                ...CharacterFragment
              }
            }
          }
        }

        fragment CharacterFragment on Character {
          name
          __typename

          ... on Human {
            homePlanet
          }

          ... on Droid {
            primaryFunction
          }
        }
      `
      );

      const source = generateSource(context, { addTypename: true });

      expect(source).toMatchSnapshot();
    });

    test('should correctly handle doubly nested fragments on interfaces', function() {
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode) {
          hero(episode: $episode) {
            name
            friendsConnection {
              friends {
                ...CharacterFragment
              }
            }
          }
        }

        fragment CharacterFragment on Character {
          name

          ... on Human {
            homePlanet
            friends {
              ...OtherCharacterFragment
            }
          }

          ... on Droid {
            primaryFunction
          }
        }

        fragment OtherCharacterFragment on Character {
          ... on Human {
            height
          }

          ... on Droid {
            appearsIn
          }
        }
      `
      );

      const source = generateSource(context);

      expect(source).toMatchSnapshot();
    });
  });
});
