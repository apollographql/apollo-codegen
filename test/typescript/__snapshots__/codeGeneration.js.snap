// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript code generation #generateSource() should correctly add typename to nested fragments on interfaces if addTypename is true 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroQueryQueryVariables {
  episode: Episode | null;
}

export interface HeroQueryQuery {
  hero: {
    // The name of the character
    name: string,
    // The friends of the character exposed as a connection with edges
    friendsConnection: {
      // A list of the friends, as a convenience when edges are not needed.
      friends: Array<CharacterFragmentFragment>,
    },
  } | null;
}

export interface CharacterFragmentFragmentOnHuman {
  __typename: \\"Human\\";
  // What this human calls themselves
  name: string;
  // The home planet of the human, or null if unknown
  homePlanet: string | null;
}

export interface CharacterFragmentFragmentOnDroid {
  __typename: \\"Droid\\";
  // What others call this droid
  name: string;
  // This droid's primary function
  primaryFunction: string | null;
}

export type CharacterFragmentFragment =
  | CharacterFragmentFragmentOnHuman
  | CharacterFragmentFragmentOnDroid
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should correctly handle doubly nested fragments on interfaces 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroQueryQueryVariables {
  episode: Episode | null;
}

export interface HeroQueryQuery {
  hero: {
    // The name of the character
    name: string,
    // The friends of the character exposed as a connection with edges
    friendsConnection: {
      // A list of the friends, as a convenience when edges are not needed.
      friends: Array<CharacterFragmentFragment>,
    },
  } | null;
}

export interface CharacterFragmentFragmentOnHuman {
  // What this human calls themselves
  name: string;
  // The home planet of the human, or null if unknown
  homePlanet: string | null;
  // This human's friends, or an empty list if they have none
  friends: Array<OtherCharacterFragmentFragment>;
}

export interface CharacterFragmentFragmentOnDroid {
  // What others call this droid
  name: string;
  // This droid's primary function
  primaryFunction: string | null;
}

export type CharacterFragmentFragment =
  | CharacterFragmentFragmentOnHuman
  | CharacterFragmentFragmentOnDroid

export interface OtherCharacterFragmentFragmentOnHuman {
  // Height in the preferred unit, default is meters
  height: number | null;
}

export interface OtherCharacterFragmentFragmentOnDroid {
  // The movies this droid appears in
  appearsIn: Array< Episode | null >;
}

export type OtherCharacterFragmentFragment =
  | OtherCharacterFragmentFragmentOnHuman
  | OtherCharacterFragmentFragmentOnDroid
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should correctly handle fragments on interfaces 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroQueryQueryVariables {
  episode: Episode | null;
}

export interface HeroQueryQuery {
  hero: HeroFragment | null;
}

export interface HeroFragmentOnHuman {
  // What this human calls themselves
  name: string;
  // The home planet of the human, or null if unknown
  homePlanet: string | null;
}

export interface HeroFragmentOnDroid {
  // What others call this droid
  name: string;
  // This droid's primary function
  primaryFunction: string | null;
}

export type HeroFragment =
  | HeroFragmentOnHuman
  | HeroFragmentOnDroid
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should correctly handle nested fragments on interfaces 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroQueryQueryVariables {
  episode: Episode | null;
}

export interface HeroQueryQuery {
  hero: {
    // The name of the character
    name: string,
    // The friends of the character exposed as a connection with edges
    friendsConnection: {
      // A list of the friends, as a convenience when edges are not needed.
      friends: Array<CharacterFragmentFragment>,
    },
  } | null;
}

export interface CharacterFragmentFragmentOnHuman {
  __typename: \\"Human\\";
  // What this human calls themselves
  name: string;
  // The home planet of the human, or null if unknown
  homePlanet: string | null;
}

export interface CharacterFragmentFragmentOnDroid {
  __typename: \\"Droid\\";
  // What others call this droid
  name: string;
  // This droid's primary function
  primaryFunction: string | null;
}

export type CharacterFragmentFragment =
  | CharacterFragmentFragmentOnHuman
  | CharacterFragmentFragmentOnDroid
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate correct typedefs with a multiple custom fragments 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroAndFriendsNamesQueryVariables {
  episode: Episode | null;
}

export interface HeroAndFriendsNamesQuery {
  hero: {
    // The name of the character
    name: string,
    // The friends of the character, or an empty list if they have none
    friends: Array<FriendFragment & PersonFragment>,
  } | null;
}

export interface FriendFragment {
  // The name of the character
  name: string;
}

export interface PersonFragment {
  // The name of the character
  name: string;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate correct typedefs with a single custom fragment 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroAndFriendsNamesQueryVariables {
  episode: Episode | null;
}

export interface HeroAndFriendsNamesQuery {
  hero: {
    // The name of the character
    name: string,
    // The friends of the character, or an empty list if they have none
    friends: Array<FriendFragment>,
  } | null;
}

export interface FriendFragment {
  // The name of the character
  name: string;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate fragmented query operations 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

export interface HeroAndFriendsNamesQuery {
  hero: HeroFriendsFragment & {
    // The name of the character
    name: string,
  } | null;
}

export interface HeroFriendsFragment {
  // The friends of the character, or an empty list if they have none
  friends: Array< {
    // The name of the character
    name: string,
  } > | null;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate mutation operations with complex input types 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface ReviewInput {
  // 0-5 stars
  stars: number;
  // Comment about the movie, optional
  commentary: string | null;
  // Favorite color, optional
  favorite_color: ColorInput | null;
}

export interface ColorInput {
  red: number;
  green: number;
  blue: number;
}

export interface ReviewMovieMutationVariables {
  episode: Episode | null;
  review: ReviewInput | null;
}

export interface ReviewMovieMutation {
  createReview: {
    // The number of stars this review gave, 1-5
    stars: number,
    // Comment about the movie
    commentary: string | null,
  } | null;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate query operations with inline fragments 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

export interface HeroAndDetailsQuery {
  hero: HeroDetailsFragment & {
    // The name of the character
    name: string,
  } | null;
}

export interface HeroDetailsFragmentOnDroid {
  // This droid's primary function
  primaryFunction: string | null;
}

export interface HeroDetailsFragmentOnHuman {
  // Height in the preferred unit, default is meters
  height: number | null;
}

export type HeroDetailsFragment =
  | HeroDetailsFragmentOnDroid
  | HeroDetailsFragmentOnHuman
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate simple nested query operations including input variables 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroAndFriendsNamesQueryVariables {
  episode: Episode | null;
}

export interface HeroAndFriendsNamesQuery {
  hero: {
    // The name of the character
    name: string,
    // The friends of the character, or an empty list if they have none
    friends: Array< {
      // The name of the character
      name: string,
    } > | null,
  } | null;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate simple query operations 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

export interface HeroNameQuery {
  hero: {
    // The name of the character
    name: string,
  } | null;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate simple query operations including input variables 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroNameQueryVariables {
  episode: Episode | null;
}

export interface HeroNameQuery {
  hero: {
    // The name of the character
    name: string,
  } | null;
}
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should handle complex fragments with type aliases 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroAndFriendsNamesQuery {
  hero: SomethingFragment & {
    // The name of the character
    name: string,
  } | null;
  empireHero: SomethingFragment & {
    // The name of the character
    name: string,
  } | null;
}

export interface SomethingFragmentOnHuman {
  // This human's friends, or an empty list if they have none
  friends: Array< FriendFragment | null > | null;
}

export interface FriendFragmentOnHuman {
  // The home planet of the human, or null if unknown
  homePlanet: string | null;
}

export interface FriendFragmentOnDroid {
  // This droid's primary function
  primaryFunction: string | null;
}

export type FriendFragment =
  | FriendFragmentOnHuman
  | FriendFragmentOnDroid

export interface SomethingFragmentOnDroid {
  // The movies this droid appears in
  appearsIn: Array< Episode | null >;
}

export type SomethingFragment =
  | SomethingFragmentOnHuman
  | SomethingFragmentOnDroid
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should handle multi-fragmented query operations 1`] = `
"//  This file was automatically generated and should not be edited.
/* tslint:disable */

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export interface HeroAndFriendsNamesQuery {
  hero: HeroFriendsFragment & HeroAppearsFragment & {
    // The name of the character
    name: string,
  } | null;
}

export interface HeroFriendsFragment {
  // The friends of the character, or an empty list if they have none
  friends: Array< {
    // The name of the character
    name: string,
  } > | null;
}

export interface HeroAppearsFragment {
  // The movies this character appears in
  appearsIn: Array< Episode | null >;
}
/* tslint:enable */
"
`;
